src/main/scala/piconot/parser/PiconotParser.scala

```diff
@@ -0,0 +1,164 @@
+package piconot.parser
+
+import scala.util.parsing.combinator._
+import piconot.ir._
+import piconot.ir.sugar._
+import picolib.semantics._
+
+object PiconotParser extends JavaTokenParsers with PackratParsers {
+    // parsing interface
+    def apply(s: String): ParseResult[AST] = {
+      parseAll(multiTransformer, s)
+    }
+    
+    def word(s: String): Parser[String] = {
+      val msg = if (s == "else") {
+        "Expected the keyword 'else'."
+      } else {
+        s"Expected the keyword '${s}'."
+      }
+      ident.filter(_ == s).withFailureMessage(msg)
+        
+    }
+    
+    
+    //lazy val ident2: PackratParser[String] = ident.filter(s => !(reserved contains s))
+    
+    def endsInBlock(trans: Transformers): Boolean = {
+      trans match {
+        case BracedTransformers(_) => true
+        case BaseTransformers(trans) => endsInBlock(trans)
+      }
+    }
+
+    def endsInBlock(trans: Transformer): Boolean = {
+      trans match {
+        case ElseTransformerBasic(_, t) => endsInBlock(t)
+        case ElseTransformerComplex(_, _, t) => endsInBlock(t)
+        case AugmentTransformer(_, t) => endsInBlock(t)
+        case BaseTransformer(_) => false
+      }
+    }
+    
+    def transformerBlockEnd: PackratParser[Transformer] = new PackratParser[Transformer] {
+      def apply(in: Input): ParseResult[Transformer] = {
+        val parse = transformer(in)
+        parse match {
+          case Success(trans, _) =>  if (endsInBlock(trans)) {
+              parse
+            } else {
+              Failure("You must end rules with either a ; or a {block}. Perhaps you forgot a ;?", in)
+            }
+          case _ => parse
+        }
+      }
+    }
+    
+    def ident2: PackratParser[String] = new PackratParser[String] {
+      def apply(in: Input): ParseResult[String] = {
+        val reserved = "else move up down left right north south east west n s e w open free blocked closed".split(" ")
+        val parse = (ident.withFailureMessage("Expected to see a state name. State names must be valid identifiers (alphanumeric, etc.)"))(in)
+        parse match {
+          case Success(word, _) => if (reserved contains word) {
```

Comment:
Fun fact: I believe Scala lets you put an `if` statement on the left side of the arrow in a `case` statement after you've pattern matched. With this method, you could drop the `else` in your first case, and let it be caught by the second.